src/plots/cartesian/axes.js

```diff
@@ -2036,6 +2037,26 @@ axes.doTicks = function(gd, axid, skipTitle) {
             });
         }
 
+        // How much to shift a multi-line label to center it vertically.
+        function getAnchorHeight(lineCount, lineHeight, angle) {
+            var h = (lineCount - 1) * lineHeight;
+            if(axLetter === 'x') {
+                if(angle < -20 || 20 < angle) {
+                    return -0.5 * h;
+                } else if(axside === 'top') {
+                    return -h;
+                }
+            } else {
+                angle *= axside === 'left' ? 1 : -1;
+                if(angle < -20) {
+                    return -h;
+                } else if(angle < 20) {
+                    return -0.5 * h;
+                }
+            }
+            return 0;
```

Comment:
Very nice, I don't think this is overly long at all, it has pretty much exactly the complexity required.

The new mock is fantastic, clearly covers all the cases we care about. The only question in my mind is the cutoff angle - why did you choose 20 degrees? It looks a little funny to me in the 45-degree cases where the x and y axis labels are essentially symmetric (look at x5/y5, and rotate the image 90 degrees CCW), but the x axis labels are vertically centered while the y axis labels are aligned to the first line. Symmetry would say the logic for an x axis at angle `angle` should match the logic for a y axis at angle `90 - angle`.

Thinking also about cases where this could cause problems: the main issue I see is if you have a hard edge on the plot (either `axis.showline` or something like bars or a heatmap flush with the edge) particularly when you're *not* showing ticks. In order for multiline labels to not overlap the plot in that case, we need to make sure the angle between the vertical text and the axis is fairly small before we vertically center the text. So that would argue for a fairly *small* cutoff in the y case (could be 20 as you have it, or maybe 30?) and a correspondingly *large* cutoff in the x case (70 or 60).

Reply:
I had no particular reason for 20 degrees. I think I ended up with the asymmetric setup because it counteracts an existing asymmetry a bit. But that's not a good reason. :slightly_smiling_face: I've switched to your recommendation. (30° cutoff for _y_, 60° cutoff for _x_.)

You can see the asymmetry I mean if you look in the top right on the current baseline. Both place the anchor at the start of the third line, but the gap between the tick and the text is greater on the top than on the right. Also the gap for `x 2` is greater than for `y 1`.

I suppose this is unrelated to this pull request though, and does not really bother my eye.